driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -518,9 +521,15 @@ private <T> void sendCommandMessageAsync(final int messageId, final Decoder<T> d
                 commandEventSender.sendSucceededEventForOneWayCommand();
                 callback.onResult(null, null);
             } else {
-                Optional<MongoOperationTimeoutException> e = validateHasTimedOutAndClose(commandEventSender, operationContext);
-                if (e.isPresent()) {
-                    callback.onResult(null, e.get());
+                boolean shouldReturn = Timeout.run(operationContext.getTimeoutContext().timeoutIncludingRoundTrip(), NANOSECONDS,
+                        () -> false,
+                        (ns) -> false,
+                        () -> {
+                            callback.onResult(null, createMongoOperationTimeoutExceptionAndClose(commandEventSender));
+                            return true;
+                        },
+                        () -> false);
+                if (shouldReturn) {
```

Comment:
Worth noting that this messy usage (and at least one other) would be avoided under the async API. It would be a throw, which would cease execution, and the exception would be handled by the  API.